Add support for pulling username from keyring subprocess provider

jfly opened the pull request:
This fixes https://github.com/pypa/pip/issues/12543.

Somewhat confusingly, when using using `PIP_KEYRING_PROVIDER=import`,
pip was able to fetch both a username and a password from keyring, but
when using `PIP_KEYRING_PROVIDER=subprocess`, it needed the username.

I had to rework the existing tests quite a bit to fit with this new
behavior, as it's now OK to ask for a username/password for an index
even if you don't have a username. This is why `KeyringSubprocessResult`
now subclasses `KeyringModuleV2`. While I was in here, I opted to remove
the "fixtures" values from `KeyringModuleV2` by introducing this new
`add_credential` contextmanager. IMO, they were hurting the readability
of our tests: you had to jump around quite a bit to see what the
contents of the keyring would be during our tests. It also forced all
our tests to play nicely with the same fixtures values, which IMO was an
unnecessary constraint.

Note: per the discussion
[here](https://github.com/pypa/pip/pull/12748#issuecomment-2294948227),
I've opted not to implement any "feature detection" to see if the
`keyring` subprocess supports `--mode=creds`. For the record, this
feature was added in
https://github.com/jaraco/keyring/commit/7830a6418a77edfe28d4c4ba3142a2a6f8a70de6,
which landed in keyring v25.2.0.

jfly committed:
Add support for pulling username from keyring subprocess provider

This fixes https://github.com/pypa/pip/issues/12543.

Somewhat confusingly, when using using `PIP_KEYRING_PROVIDER=import`,
pip was able to fetch both a username and a password from keyring, but
when using `PIP_KEYRING_PROVIDER=subprocess`, it needed the username.

I had to rework the existing tests quite a bit to fit with this new
behavior, as it's now OK to ask for a username/password for an index
even if you don't have a username. This is why `KeyringSubprocessResult`
now subclasses `KeyringModuleV2`. While I was in here, I opted to remove
the "fixtures" values from `KeyringModuleV2` by introducing this new
`add_credential` contextmanager. IMO, they were hurting the readability
of our tests: you had to jump around quite a bit to see what the
contents of the keyring would be during our tests. It also forced all
our tests to play nicely with the same fixtures values, which IMO was an
unnecessary constraint.

Note: per the discussion
[here](https://github.com/pypa/pip/pull/12748#issuecomment-2294948227),
I've opted not to implement any "feature detection" to see if the
`keyring` subprocess supports `--mode=creds`. For the record, this
feature was added in
https://github.com/jaraco/keyring/commit/7830a6418a77edfe28d4c4ba3142a2a6f8a70de6,
which landed in keyring v25.2.0.

jfly commented:
@Darsstar, I would love your feedback on this, as you seem to be the primary maintainer of pip's keyring support.

Darsstar commented:
'Maintainer' is more than I am willing to claim, but I do keep an eye on the feature I contributed to, hopefully, ease the maintenance on it.

I could personally live with the breaking change. But feature detection would be nice.

The `getcreds` subcommand is actually even more recent, v25.2.1, so please don't use that for feature detection.

jfly commented:
> The getcreds subcommand is actually even more recent, v25.2.1

Ok, so I was a little confused. But I think you are too? It looks like https://github.com/jaraco/keyring/pull/678 went through some iteration before getting merged up, and all the commits got included in the merge:

1. `getcreds` was introduced in https://github.com/jaraco/keyring/commit/0a104c88afa04e22906654bc6d7dbd8b19a012e9 
2. `getcreds` was removed and replaced with `--mode=creds` in https://github.com/jaraco/keyring/commit/7830a6418a77edfe28d4c4ba3142a2a6f8a70de6

Both commits have landed in keyring v25.2.0+, but commit 1 (the "getcreds" command) never actually landed in a release without commit 2 (the `--mode=creds`) command.

> I could personally live with the breaking change. But feature detection would be nice. [...] please don't use that [getcreds] for feature detection.

Agreed. I guess I'll wait for feedback from a pip maintainer before I got to the effort of implementing feature detection.

Darsstar commented:
> But I think you are too?

Yeah... I would have sworn I saw a newer commit adding it 'back' after 25.2.0...

jfly commented:
For the record, I'm aware of the test failures in `test_keyring_cli_get_password`. I can fix them, but would like to get feedback on the "do we need feature detection of the keyring subprocess?" question before I do, as it'll affect how I update that test to pass.

jfly commented:
> Hello!
> 
> I am an automated bot and I have noticed that this pull request is not currently able to be merged. If you are able to either merge the `master` branch into this pull request or rebase this pull request against `master` then it will eligible for code review and hopefully merging!

Oops, I accidentally pushed a WIP commit to fix the tests.

Again, I'd like to get some feedback on the "do we need feature detection of the keyring subprocess?" question before I finalize this work, as it'll affect how I update that test to pass.

@Darsstar are you the right person to talk to about this?

Darsstar commented:
No, I'm just a user. A user who is fine with the minor churn no feature detection would cause.

jfly commented on tests/unit/test_network_auth.py:
```diff
@@ -443,31 +552,42 @@ def check_returncode(self) -> None:
 
-    actual = auth._get_new_credentials(url, allow_netrc=False, allow_keyring=True)
-    assert actual == expect
+    with keyring_subprocess.add_credential("example.com", "example", "!netloc"):
+        with keyring_subprocess.add_credential(
+            "http://example.com/path2/", "saved-user1", "pw1"
+        ):
+            with keyring_subprocess.add_credential(
+                "http://example.com/path2/", "saved-user2", "pw2"
+            ):
```
Urg, so this got ugly. I was hoping to use this nicer syntax for opening multiple context managers at once, but apparently it's not supported on Python 3.8:

```suggestion
    with (
        keyring_subprocess.add_credential("example.com", "example", "!netloc"),
        keyring_subprocess.add_credential("http://example.com/path2/", "saved-user1", "pw1"),
        keyring_subprocess.add_credential("http://example.com/path2/", "saved-user2", "pw2"),
    ):
```

I could rework this `add_credential` method to instead take an array of triplets or something, but IMO that makes the api kind of messy.

I'm inclined to leave this the way it is (and clean it up when we drop support for versions of python that don't support the above syntax), but I'm happy to approach this however folks prefer.

jfly committed:
Address test failures in tests/functional/test_install_config.py

I found that with my changes, some of the xfails in here started failing
(because the underlying tests are actually passing now). I found the
existing test very difficult to understand (I'm honestly not sure I
undertood them, especially the logic around `keyring_provider` in the
test setup), so I opted to refactor things a bit to hopefully make them
make sense.

jfly commented on tests/functional/test_install_config.py:
```diff
@@ -416,22 +417,32 @@ def test_prompt_for_keyring_if_needed(
-    script = script_factory(workspace.joinpath("venv"), virtualenv, environ=environ)
-
-    if (
-        keyring_provider not in [None, "auto"]
```
I found the pre-existing code hard to grok. This line in particular really confounds me: why does `keyring_provider` influence how we set up the test environment? I'd expect that to be only dependent upon `keyring_provider_implementation` (which is the change I've made in this PR).

jfly replied:
> After your refactoring keyring won't be importable when keyring_provider in [None, "auto"] and keyring_provider_implementation == "disabled".

Yes, this was intentional. Perhaps I'm misunderstanding these variable names? I read them as:

- `keyring_provider_implementation`: this is the way i'm going to set up my environment for this test (is keyring installed in this venv, is is discoverable via PATH in a different venv, or is it not installed at all)
- `keyring_provider`: this is the keyring provider i'm going to request when i invoke pip

So, it's intentional that if `keyring_provider_implementation == "disabled"`, then yeah, keyring won't be imprtable.

But again, it's very possible I'm misunderstanding the intent of these tests. Guidance appreciated.

> Maybe it is an idea to ask for verbose output and check for the Keyring provider set: lines to make that more explicit?

Yeah, I personally not in love with this elaborate usage of parameterize + xfail. Tests can fail for all sorts of reasons other than the expected reason! I think it would be better to assert on the expected failures. (IMO, that would be better to do in a separate PR)

jfly commented:
I've completed the implementation and have gotten the tests passing.

@Darsstar if you're up for giving me a review, I'd sure appreciate it!

YakBizzarro commented:
hey, I bumped on this bug and saw this fix.
I know a lot of time passed, but is there still a chance this gets merged?
If the original author is not interested anymore, would be possible to "adopt" this PR? In the case, what would be needed to do? Most of the issue I see in the discussion are probably not applicable anymore due to the long time passed. Keyring 25.2.0 has been out for two years and Python 3.8 is not supported anymore
Many thanks!

sepehr-rs commented:
Hi @YakBizzarro
The PR currently has merge conflicts, and discussion stalled after the last round of feedback. I don't see any maintainer objection to the feature itself, so I think it's something pip would welcome; it just needs someone to resolve the conflicts and pick up the remaining open question in the thread to move it forward.

jfly commented:
I received some unenthusiastic feedback from on a related keyring PR: https://github.com/pypa/pip/pull/12934#issuecomment-2585060853, which made it seem like pinging people to review *this* PR would be unwelcome.

Perhaps things have changed in the intervening time, though.

I'm (happily) no longer dealing with AWS CodeArtifact, so I don't have the same motivation to push this over the finish line.

If somebody wants to adopt this PR, feel free. I'd just ask that you preserve attribution with a Co-authored-by trailer.

YakBizzarro commented:
Thanks @sepehr-rs and @jfly ! I opened a new PR at https://github.com/pypa/pip/pull/14150
